### ISSUE TITLE
Proper Dead-code Elimination

### ISSUE THREAD

**Issue by thomasrolinger** [thomasrolinger/chapel#41]
The method we use now for the static checks does not actually do dead code elimination. If an optimization is invalid, we basically make the `if` check say `if false == true` but that does not remove the `then` branch.

I have a way to do this manually and it does work for many cases for the IE optimization. But it does cause issues in some cases, like if the `forall` had `+reduce` clauses. Seems that we are not quite removing things correctly, specifically during prefolding.

For aggregation, I think we can do this the "proper" way because the things we check for validity are things that Chapel code can tell us (i.e., if the array is distributed and if the aggregator is `nil`). I know the is-distributed check is `param` based but not sure about the aggregator. We certainly know if it is `nil` by prefolding.

This is not a pressing issue but it would be nice to get it working at some point.

**Comment by thomasrolinger** [comment]
I was able to do the "proper" dead code elimination for aggregation. I created a `param` procedure that returns true if the array `A` is distributed and the aggregator was created (not `nil`). Using the return of this procedure as a guard around the primitive, the primitive will only survive if the guard is resolved to true. In this way, we could just completely ignore prefolding all together, since we would generate the aggregation call within the `then` branch of the conditional. But for now, just keep prefolding so it at least provides logging to the terminal that we succeeded with the optimization. However, it also means there is no explicit printing that we did not succeed, since the primitive is gone if the `param` check is false. But that is just something we'll live with. I also fixed a bug in the IE optimization when looking for field accesses during prefolding. Issue was that we were finding unresolved versions of the executor loop since we were just matching on `forall` ID rather than something more specific to the primitive.

Here is the commit for that: https://github.com/thomasrolinger/chapel/commit/09475d84873b201f48c5d610700d673bd8658e63

We should also be able to do the same thing for prefetching.

**Comment by thomasrolinger** [comment]
For the IE optimization, the issue that the `AggregateType` that refers to one of the `+ reduce` statements has an instantiation point set within the code we eliminated. I suspect it was set in an early stage of resolution, so it's a done-deal by the time we get the our prefolding removal code.

I am working on an approach that will just postpone any prefolding removal until post resolution. I tested that this does seem to work just fine. So the idea is to somehow log which `staticCheckCall` `CallExprs` we know we want to remove (i.e., the `then` block that corresponds to the `CondStmt` that uses that `statichCheckCall` result). We could do this via the `forall->optInfo` or perhaps a global vector. Let's try the global vector approach.
